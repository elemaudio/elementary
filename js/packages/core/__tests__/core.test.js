import {
  createNode,
  renderWithDelegate,
  stepGarbageCollector,
  resolve,
  Renderer,
} from '..';


class TestRenderer extends Renderer {
  constructor() {
    super(() => {});
  }

  render(...args) {
    this._delegate.clear();

    renderWithDelegate(this._delegate, args.map(resolve));

    return {
      nodesAdded: this._delegate.nodesAdded,
      edgesAdded: this._delegate.edgesAdded,
      propsWritten: this._delegate.propsWritten,
      elapsedTimeMs: 0,
    };
  }

  getBatch() {
    return this._delegate.getPackedInstructions();
  }

  getTerminalGeneration() {
    return this._delegate.getTerminalGeneration();
  }

  gc() {
    stepGarbageCollector(this._delegate);
  }
}

function sortInstructionBatch(x) {
  let copy = [...x];

  // This is an odd sort step; in particular, we only sort here within each
  // type of instruction. That is, createNodes all come before deleteNodes, all come before
  // appendChilds, etc, in the snapshot files. The Renderer instance, via its getPackedInstructions
  // method, provides that group-level sort during render. However, the snapshots then sort
  // within those groups according to hash values so that our tests are not coupled to the traversal
  // order of the reconciler. This whole thing is temporary anyway; once we finish the v3 refactor,
  // we can remove this sorting and update the snapshots.
  return copy.sort((a, b) => {
    if (a[0] === b[0]) {
      return a[1] < b[1] ? -1 : 1;
    }

    return 0;
  });
}

test('the basics', function() {
  let tr = new TestRenderer();

  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        createNode("const", {value: 2 * Math.PI}, []),
        createNode("phasor", {}, [
          createNode("const", {key: 'fq', value: 440}, [])
        ])
      ])
    ])
  );

  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('numeric literals', function() {
  let tr = new TestRenderer();

  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        2 * Math.PI,
        createNode("phasor", {}, [440])
      ])
    ])
  );

  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('distinguish by props', function() {
  let tr = new TestRenderer();

  const renderVoice = (path, seq) => {
    return (
      createNode("sample", {path}, [
        createNode("seq", {seq}, [
          createNode("le", {}, [
            createNode("phasor", {}, [
              createNode("const", {value: 2}, [])
            ]),
            createNode("const", {value: 0.5}, [])
          ]),
        ]),
      ])
    );
  };

  tr.render(
    renderVoice('test/path.wav', [0, 0, 1]),
    renderVoice('test/path.wav', [0, 1, 0]),
  );

  // Here we haven't specified any keys, so the renderer should do a lot of
  // structural sharing. However, because we're rendering a different sequence,
  // we expect to see the pulse train (starting at the "le" node) shared, but
  // then see two different root nodes, two different sample nodes, and two
  // different seq nodes.
  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('multi-channel basics', function() {
  const tr = new TestRenderer();

  const monoProcess = (
    createNode("sin", {}, [
      createNode("mul", {}, [
        createNode("const", {value: 2 * Math.PI}, []),
        createNode("phasor", {}, [
          createNode("const", {key: 'fq', value: 440}, [])
        ])
      ])
    ])
  );

  // Run the same thing in the left and the right channel
  tr.render(monoProcess, monoProcess);

  // We should see structural sharing except for the root nodes
  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('simple sharing', function() {
  let tr = new TestRenderer();

  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        createNode("const", {value: 2 * Math.PI}, []),
        createNode("phasor", {}, [
          createNode("const", {key: 'fq', value: 440}, [])
        ])
      ])
    ])
  );

  // Second render inserts a tanh at the top, we should find this and
  // share all of the child nodes during this render pass
  tr.render(
    createNode("tanh", {}, [
      createNode("sin", {}, [
        createNode("mul", {}, [
          createNode("const", {value: 2 * Math.PI}, []),
          createNode("phasor", {}, [
            createNode("const", {key: 'fq', value: 440}, [])
          ])
        ])
      ])
    ])
  );

  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('distinguished subtrees by key', function() {
  let tr = new TestRenderer();

  const voices = [
    {key: 'fq1', freq: 440},
    {key: 'fq2', freq: 440},
    {key: 'fq3', freq: 440},
    {key: 'fq4', freq: 440},
  ];

  const renderVoice = (voice) => {
    return (
      createNode("sin", {}, [
        createNode("mul", {}, [
          createNode("const", {value: 2 * Math.PI}, []),
          createNode("phasor", {}, [
            createNode("const", {key: voice.key, value: voice.freq}, [])
          ])
        ])
      ])
    );
  };

  tr.render(createNode("add", {}, voices.map(renderVoice)))

  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('structural equality with value change', function() {
  const tr = new TestRenderer();

  const voices = [
    {key: 'fq1', freq: 440},
    {key: 'fq2', freq: 440},
    {key: 'fq3', freq: 440},
    {key: 'fq4', freq: 440},
  ];

  const renderVoice = (voice) => {
    return (
      createNode("sin", {}, [
        createNode("mul", {}, [
          createNode("const", {value: 2 * Math.PI}, []),
          createNode("phasor", {}, [
            createNode("const", {key: voice.key, value: voice.freq}, [])
          ])
        ])
      ])
    );
  };

  tr.render(createNode("add", {}, voices.map(renderVoice)))

  // Change one of the keyed values, should see structural equality
  // with no new nodes created
  voices[0].freq = 441;

  tr.render(createNode("add", {}, voices.map(renderVoice)))

  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

// Testing here to ensure that the root node activation works as expected.
test('switch and switch back', function() {
  const tr = new TestRenderer();

  const renderVoice = (voice) => {
    return (
      createNode("sin", {}, [
        createNode("mul", {}, [
          createNode("const", {value: 2 * Math.PI}, []),
          createNode("phasor", {}, [
            createNode("const", {key: voice.key, value: voice.freq}, [])
          ])
        ])
      ])
    );
  };

  // We're going to render three times with graphs A, B, A. We want to see on
  // the third render that almost nothing was done because we still have the nodes
  // around that we need, i.e. they haven't been garbage collected, but we will need
  // to see that the correct root nodes are in place.
  tr.render(renderVoice({key: 'hi', freq: 440}));
  tr.render(renderVoice({key: 'bye', freq: 880}));

  tr.render(renderVoice({key: 'hi', freq: 440}));
  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('garbage collection', function() {
  let tr = new TestRenderer();

  // First, render a sine tone with `sin`
  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        2 * Math.PI,
        createNode("phasor", {}, [440])
      ])
    ])
  );

  // Next we're going to render a sine tone using cosine. We'll expect to
  // see most of the prior structure preserved.
  tr.gc();

  tr.render(
    createNode("cos", {}, [
      createNode("mul", {}, [
        2 * Math.PI,
        createNode("phasor", {}, [440])
      ])
    ])
  );

  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
  tr._delegate.clear();

  // Now if we step the garbage collector enough we should see the sine node
  // and its parent root get cleaned up now that they're no longer referenced
  // in the active tree
  for (let i = 0; i < tr.getTerminalGeneration() - 1; ++i) {
    tr.gc();
  }

  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('refs', function() {
  let tr = new TestRenderer();

  // Sine tone with a frequency set by ref
  let [freq, setFreq] = tr.createRef("const", {value: 440}, []);

  // Render the ref
  tr.render(
    createNode("sin", {}, [
      createNode("mul", {}, [
        2 * Math.PI,
        createNode("phasor", {}, [freq])
      ])
    ])
  );

  // Using our ref setter
  setFreq({value: 550});

  // We expect here to see a single set property instruction
  // followed by the commit instruction
  expect(tr.getBatch()).toEqual([
    [ 3, 1915043800, 'value', 550 ],
    [ 5 ],
  ]);
});
