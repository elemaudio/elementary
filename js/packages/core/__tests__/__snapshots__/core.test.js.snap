// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`distinguish by props 1`] = `
[
  [
    0,
    120667947,
    "sample",
  ],
  [
    0,
    949666729,
    "seq",
  ],
  [
    0,
    998409329,
    "phasor",
  ],
  [
    0,
    1181055054,
    "const",
  ],
  [
    0,
    1229809873,
    "seq",
  ],
  [
    0,
    1423379257,
    "root",
  ],
  [
    0,
    1497369747,
    "sample",
  ],
  [
    0,
    1557538697,
    "const",
  ],
  [
    0,
    1737893648,
    "root",
  ],
  [
    0,
    2118771019,
    "le",
  ],
  [
    2,
    120667947,
    949666729,
    0,
  ],
  [
    2,
    949666729,
    2118771019,
    0,
  ],
  [
    2,
    998409329,
    1557538697,
    0,
  ],
  [
    2,
    1229809873,
    2118771019,
    0,
  ],
  [
    2,
    1423379257,
    120667947,
    0,
  ],
  [
    2,
    1497369747,
    1229809873,
    0,
  ],
  [
    2,
    1737893648,
    1497369747,
    0,
  ],
  [
    2,
    2118771019,
    998409329,
    0,
  ],
  [
    2,
    2118771019,
    1181055054,
    0,
  ],
  [
    3,
    120667947,
    "path",
    "test/path.wav",
  ],
  [
    3,
    949666729,
    "seq",
    [
      0,
      0,
      1,
    ],
  ],
  [
    3,
    1181055054,
    "value",
    0.5,
  ],
  [
    3,
    1229809873,
    "seq",
    [
      0,
      1,
      0,
    ],
  ],
  [
    3,
    1423379257,
    "channel",
    0,
  ],
  [
    3,
    1497369747,
    "path",
    "test/path.wav",
  ],
  [
    3,
    1557538697,
    "value",
    2,
  ],
  [
    3,
    1737893648,
    "channel",
    1,
  ],
  [
    4,
    [
      1423379257,
      1737893648,
    ],
  ],
  [
    5,
  ],
]
`;

exports[`distinguished subtrees by key 1`] = `
[
  [
    0,
    371590287,
    "mul",
  ],
  [
    0,
    472261513,
    "phasor",
  ],
  [
    0,
    549707649,
    "mul",
  ],
  [
    0,
    555467758,
    "const",
  ],
  [
    0,
    650660620,
    "mul",
  ],
  [
    0,
    765455487,
    "sin",
  ],
  [
    0,
    1007055215,
    "phasor",
  ],
  [
    0,
    1057677500,
    "phasor",
  ],
  [
    0,
    1097289128,
    "const",
  ],
  [
    0,
    1193164375,
    "const",
  ],
  [
    0,
    1427558172,
    "const",
  ],
  [
    0,
    1478357524,
    "sin",
  ],
  [
    0,
    1609650518,
    "sin",
  ],
  [
    0,
    1661951969,
    "const",
  ],
  [
    0,
    1802191769,
    "sin",
  ],
  [
    0,
    1875532204,
    "add",
  ],
  [
    0,
    1891960130,
    "root",
  ],
  [
    0,
    1933267990,
    "mul",
  ],
  [
    0,
    2030160862,
    "phasor",
  ],
  [
    2,
    371590287,
    1097289128,
    0,
  ],
  [
    2,
    371590287,
    2030160862,
    0,
  ],
  [
    2,
    472261513,
    1661951969,
    0,
  ],
  [
    2,
    549707649,
    1097289128,
    0,
  ],
  [
    2,
    549707649,
    1057677500,
    0,
  ],
  [
    2,
    650660620,
    1097289128,
    0,
  ],
  [
    2,
    650660620,
    472261513,
    0,
  ],
  [
    2,
    765455487,
    650660620,
    0,
  ],
  [
    2,
    1007055215,
    1193164375,
    0,
  ],
  [
    2,
    1057677500,
    1427558172,
    0,
  ],
  [
    2,
    1478357524,
    371590287,
    0,
  ],
  [
    2,
    1609650518,
    549707649,
    0,
  ],
  [
    2,
    1802191769,
    1933267990,
    0,
  ],
  [
    2,
    1875532204,
    1609650518,
    0,
  ],
  [
    2,
    1875532204,
    1802191769,
    0,
  ],
  [
    2,
    1875532204,
    1478357524,
    0,
  ],
  [
    2,
    1875532204,
    765455487,
    0,
  ],
  [
    2,
    1891960130,
    1875532204,
    0,
  ],
  [
    2,
    1933267990,
    1097289128,
    0,
  ],
  [
    2,
    1933267990,
    1007055215,
    0,
  ],
  [
    2,
    2030160862,
    555467758,
    0,
  ],
  [
    3,
    555467758,
    "key",
    "fq3",
  ],
  [
    3,
    555467758,
    "value",
    440,
  ],
  [
    3,
    1097289128,
    "value",
    6.283185307179586,
  ],
  [
    3,
    1193164375,
    "key",
    "fq2",
  ],
  [
    3,
    1193164375,
    "value",
    440,
  ],
  [
    3,
    1427558172,
    "key",
    "fq1",
  ],
  [
    3,
    1427558172,
    "value",
    440,
  ],
  [
    3,
    1661951969,
    "key",
    "fq4",
  ],
  [
    3,
    1661951969,
    "value",
    440,
  ],
  [
    3,
    1891960130,
    "channel",
    0,
  ],
  [
    4,
    [
      1891960130,
    ],
  ],
  [
    5,
  ],
]
`;

exports[`garbage collection 1`] = `
[
  [
    0,
    19075110,
    "cos",
  ],
  [
    0,
    1283161248,
    "root",
  ],
  [
    2,
    19075110,
    1131011062,
    0,
  ],
  [
    2,
    1283161248,
    19075110,
    0,
  ],
  [
    3,
    1283161248,
    "channel",
    0,
  ],
  [
    4,
    [
      1283161248,
    ],
  ],
  [
    5,
  ],
]
`;

exports[`garbage collection 2`] = `
[
  [
    5,
  ],
  [
    1,
    1151416953,
  ],
  [
    1,
    1468062699,
  ],
]
`;

exports[`multi-channel basics 1`] = `
[
  [
    0,
    704782214,
    "root",
  ],
  [
    0,
    897687670,
    "mul",
  ],
  [
    0,
    912820239,
    "phasor",
  ],
  [
    0,
    1097289128,
    "const",
  ],
  [
    0,
    1709217017,
    "sin",
  ],
  [
    0,
    1709345783,
    "const",
  ],
  [
    0,
    2134855787,
    "root",
  ],
  [
    2,
    704782214,
    1709217017,
    0,
  ],
  [
    2,
    897687670,
    1097289128,
    0,
  ],
  [
    2,
    897687670,
    912820239,
    0,
  ],
  [
    2,
    912820239,
    1709345783,
    0,
  ],
  [
    2,
    1709217017,
    897687670,
    0,
  ],
  [
    2,
    2134855787,
    1709217017,
    0,
  ],
  [
    3,
    704782214,
    "channel",
    1,
  ],
  [
    3,
    1097289128,
    "value",
    6.283185307179586,
  ],
  [
    3,
    1709345783,
    "key",
    "fq",
  ],
  [
    3,
    1709345783,
    "value",
    440,
  ],
  [
    3,
    2134855787,
    "channel",
    0,
  ],
  [
    4,
    [
      2134855787,
      704782214,
    ],
  ],
  [
    5,
  ],
]
`;

exports[`numeric literals 1`] = `
[
  [
    0,
    198353271,
    "const",
  ],
  [
    0,
    1097289128,
    "const",
  ],
  [
    0,
    1131011062,
    "mul",
  ],
  [
    0,
    1151416953,
    "sin",
  ],
  [
    0,
    1468062699,
    "root",
  ],
  [
    0,
    1734917263,
    "phasor",
  ],
  [
    2,
    1131011062,
    1097289128,
    0,
  ],
  [
    2,
    1131011062,
    1734917263,
    0,
  ],
  [
    2,
    1151416953,
    1131011062,
    0,
  ],
  [
    2,
    1468062699,
    1151416953,
    0,
  ],
  [
    2,
    1734917263,
    198353271,
    0,
  ],
  [
    3,
    198353271,
    "value",
    440,
  ],
  [
    3,
    1097289128,
    "value",
    6.283185307179586,
  ],
  [
    3,
    1468062699,
    "channel",
    0,
  ],
  [
    4,
    [
      1468062699,
    ],
  ],
  [
    5,
  ],
]
`;

exports[`simple sharing 1`] = `
[
  [
    0,
    1442301653,
    "tanh",
  ],
  [
    0,
    1832877767,
    "root",
  ],
  [
    2,
    1442301653,
    1709217017,
    0,
  ],
  [
    2,
    1832877767,
    1442301653,
    0,
  ],
  [
    3,
    1832877767,
    "channel",
    0,
  ],
  [
    4,
    [
      1832877767,
    ],
  ],
  [
    5,
  ],
]
`;

exports[`structural equality with value change 1`] = `
[
  [
    3,
    1427558172,
    "value",
    441,
  ],
  [
    5,
  ],
]
`;

exports[`switch and switch back 1`] = `
[
  [
    4,
    [
      2063899281,
    ],
  ],
  [
    5,
  ],
]
`;

exports[`the basics 1`] = `
[
  [
    0,
    897687670,
    "mul",
  ],
  [
    0,
    912820239,
    "phasor",
  ],
  [
    0,
    1097289128,
    "const",
  ],
  [
    0,
    1709217017,
    "sin",
  ],
  [
    0,
    1709345783,
    "const",
  ],
  [
    0,
    2134855787,
    "root",
  ],
  [
    2,
    897687670,
    1097289128,
    0,
  ],
  [
    2,
    897687670,
    912820239,
    0,
  ],
  [
    2,
    912820239,
    1709345783,
    0,
  ],
  [
    2,
    1709217017,
    897687670,
    0,
  ],
  [
    2,
    2134855787,
    1709217017,
    0,
  ],
  [
    3,
    1097289128,
    "value",
    6.283185307179586,
  ],
  [
    3,
    1709345783,
    "key",
    "fq",
  ],
  [
    3,
    1709345783,
    "value",
    440,
  ],
  [
    3,
    2134855787,
    "channel",
    0,
  ],
  [
    4,
    [
      2134855787,
    ],
  ],
  [
    5,
  ],
]
`;
